Create Hypothetical Data

Eric-Sommer opened the pull request:
### What problem do you want to solve?

closes #26 
Following @amageh's remarks on creating sensible data for the tutorial, I recycled this bit to create a dataset which can be fed into gettsim. 

Includes:

- six standard household types:

|             | Single  | Couple  |
|-------------|---------|---------|
| no children | *single_0_children*  | *couple_0_children*  | 
| 1 child     | *single_1_children* | *couple_1_children* | 
| 2 children  | *single_2_children* | *couple_2_children* |

- `bedarfsgemeinschaften.yaml` is a reference for realistic rent and dwelling size values from statistics on SGB II recipients. This is important for analyzing transitions out of the benefit system.

For example, this bit in `basic_interface.ipynb`:

```python
data = pd.DataFrame(
    {
        "p_id": 1,
        "tu_id": 1,
        "hh_id": 1,
        "bruttolohn_m": 2000,
        "wohnort_ost": False,
        "alter": 35,
        "selbstständig": False,
        "hat_kinder": False,
        "eink_selbst_m": 0,
        "ges_rente_m": 0,
        "prv_krankenv": False,
    },
    index=[0],
)
```
can be replaced by 
```python
df = create_synthetic_date(hh_typ=["sing"], bruttolohn=2000)
```

The function currently has a limited number of arguments. It can be easily extended for e.g. creating self-employed, pensioners, households with more children etc. It should cover the 'basic needs', though.


Example Output:

![image](https://user-images.githubusercontent.com/44068467/97790456-fc958a80-1bc8-11eb-8a4a-765e64b1a5a0.png)


### Todo

- [x] Import the parameters into `gettsim_hypo_data()`, similar to loading tax-benefit parameters.
- [x] implement this framework in the tutorials
- [x] implement some basic tests
- [x] run synthetic data through gettsim

- Comments by @amageh:
- [x] rename to `create_synthetic_data`
- [x] `hh_typ` and `n_children` are distinct arguments
- [x] arguments for setting income and wealth now in line with column names
- [x] specify household members via lists

**optional**   
- [ ] include a targets argument synonymous to `compute_taxes_and_transfers` where users could specify a target or list of targets they want to compute, and the function generates data containing all necessary input columns for the target(s).
- [x] introduce more heterogeneity in incomes and wealth.
  - [x] **partly done**, unique household id's still left to do

amageh commented:
Hi! I think this looks really neat and would generally be very useful for the tutorials and/or new users. I am not sure whether I should review though @ChristianZimpelmann ? I am happy to leave some comments if you guys want.

hmgaudecker commented:
> Hi! I think this looks really neat and would generally be very useful for the tutorials and new users. I am not sure whether I should review though @ChristianZimpelmann ? I am happy to leave some comments if you guys want.

We would be very happy about your comments, I am sure! But feel free to decline the request, after all you have moved on to greener pastures in your daily work :wink: 

In any case, I just merged #232 and the first step before a review would be to rebase on master, @Eric-Sommer. Thanks for your work here, looks great!

ChristianZimpelmann commented:
> > Hi! I think this looks really neat and would generally be very useful for the tutorials and new users. I am not sure whether I should review though @ChristianZimpelmann ? I am happy to leave some comments if you guys want.
> 
> We would be very happy about your comments, I am sure! But feel free to decline the request, after all you have moved on to greener pastures in your daily work wink
> 
> In any case, I just merged #232 and the first step before a review would be to rebase on master, @Eric-Sommer. Thanks for your work here, looks great!

What Hans-Martin said. I will also have a look at it, but your comments would be helpful for sure.

amageh commented:
Alright, thanks, then these are some thoughts I have :-) :

- I would probably rename the function `gettsim_hypo_data` (and all associated names similarly) to `simulate_data` / `generate_data` or `get_synthetic_data`/ `get_artifical_data` . The current name is not as descriptive as it could be in my opinion.
- Similarly, I would rename the `hh_typen` inputs to: `single`, `single_one_child`, `single_two_child`, `couple`, `couple_one_child`, `couple_two_child` or something like this because these would more self-explanatory (if there is a naming convention I am not aware of or a reason to save on letters in function inputs, just ignore this comment).
- As far as I am aware, arguments of user-facing functions are usually in English. If I interpret the GEP on naming conventions correctly, German is used if it reflects a specific law. I would either name the arguments according to specific input names (i.e. `bruttolohn_m` instead of `bruttolohn`) or use English names (since using a specific input name for instance doesn't work for `alter` with the current structure).
- Maybe it would be easier to specify household members via lists (i.e. instead of having `alter_kind_1` and `alter_kind_2` as arguments, users could just pass a list like `age_child = [3, 8]`. I would do it the same way for adults for consistency.
- I didn't review the code because I assume it is not done yet, but in the current version, there seems to be a bug in `hypo.py` line 43-46 which sets `bruttolohn_m` to 0 (I assume the DataFrame there should be called `df_new` instead of `df`).


I also had some more extensive ideas while working on the tutorials. Not sure if these are relevant for the use-cases you want to cover, I am just using the chance to throw them out there:

- If it wouldn't be too difficult to implement, I think a really useful feature of this function could be if it had a `targets` argument synonymous to `compute_taxes_and_transfers` where users could specify a target or list of targets they want to compute, and the function generates data containing all necessary input columns for the target(s). I think this would be very useful because most new users will likely struggle with identifying all the inputs they need for a certain target. The error messages in `compute_taxes_and_transfers` already cover this issue but adding such a feature to the data simulator would have the added benefit of showing users exactly how the data needs to look like.
- It would be quite cool to introduce some more heterogeneity to the data (for instance in wages). Introducing this for only selected variables like `bruttolohn_m` wouldn't be too difficult I assume. Another approach could be implementing some sort of more abstract sampling rules that work for all input variables (maybe depending on the type of the variable). For this, users could maybe pass a dictionary where the key corresponds to the input which shall be sampled and the value specifies how it should be sampled. This is just an idea though, might be too complex to implement or even use and go way beyond what the function is needed for.

Eric-Sommer commented:
>     * I would probably rename the function `gettsim_hypo_data` (and all associated names similarly) to `simulate_data` / `generate_data` or `get_synthetic_data`/ `get_artifical_data` . The current name is not as descriptive as it could be in my opinion.
> 
>     * Similarly, I would rename the `hh_typen` inputs to: `single`, `single_one_child`, `single_two_child`, `couple`, `couple_one_child`, `couple_two_child` or something like this because these would more self-explanatory (if there is a naming convention I am not aware of or a reason to save on letters in function inputs, just ignore this comment).
> 
>     * As far as I am aware, arguments of user-facing functions are usually in English. If I interpret the GEP on naming conventions correctly, German is used if it reflects a specific law. I would either name the arguments according to specific input names (i.e. `bruttolohn_m` instead of `bruttolohn`) or use English names (since using a specific input name for instance doesn't work for `alter` with the current structure).
> 
>     * Maybe it would be easier to specify household members via lists (i.e. instead of having `alter_kind_1` and `alter_kind_2` as arguments, users could just pass a list like `age_child = [3, 8]`. I would do it the same way for adults for consistency.

Makes all sense to me!
 
>     * I didn't review the code because I assume it is not done yet, but in the current version, there seems to be a bug in `hypo.py` line 43-46 which sets `bruttolohn_m` to 0 (I assume the DataFrame there should be called `df_new` instead of `df`).

Correct, I fixed that one, thanks! Don't hesitate to review the code.

>     * If it wouldn't be too difficult to implement, I think a really useful feature of this function could be if it had a `targets` argument synonymous to `compute_taxes_and_transfers` where users could specify a target or list of targets they want to compute, and the function generates data containing all necessary input columns for the target(s). I think this would be very useful because most new users will likely struggle with identifying all the inputs they need for a certain target. The error messages in `compute_taxes_and_transfers` already cover this issue but adding such a feature to the data simulator would have the added benefit of showing users exactly how the data needs to look like.
 
That's certainly doable, but I haven't had any experience yet with the new framework.

>     * It would be quite cool to introduce some more heterogeneity to the data (for instance in wages). Introducing this for only selected variables like `bruttolohn_m` wouldn't be too difficult I assume. Another approach could be implementing some sort of more abstract sampling rules that work for all input variables (maybe depending on the type of the variable). For this, users could maybe pass a dictionary where the key corresponds to the input which shall be sampled and the value specifies how it should be sampled. This is just an idea though, might be too complex to implement or even use and go way beyond what the function is needed for.

That should be indeed relatively easy. I'll give it a try. The easiest way is probably to loop around the `gettsim_hypo_data()` function. We should

ChristianZimpelmann reviewed (commented):
@Eric-Sommer, thank you for the initiative!

I agree with Annica's points and added some additional remarks.

Would you like to continue working on this? If yes, the next steps would be, in my opinion, to resolve our comments and add two issues for Annica's future ideas. No need to address those now.

Eric-Sommer commented:
> Would you like to continue working on this? If yes, the next steps would be, in my opinion, to resolve our comments and add two issues for Annica's future ideas. No need to address those now.

Yes I'd like :) I actually addressed a few of Anica's points already, including the heterogeneity issue. I'll track those in the first comment.

Eric-Sommer commented on gettsim/syntethic.py:
```diff
@@ -0,0 +1,203 @@
+
+
+def gettsim_hypo_data(
+    hh_typen=("sing", "sp1ch", "sp2ch", "coup", "coup1ch", "coup2ch"),
```
one should rather use a list instead of a tuple (the docstring was wrong about that). The reason the default argument is specified as a tuple is flake8 which does not accept mutable default arguments.

ChristianZimpelmann replied:
Ok, makes sense.

I would suggest `hh_typen=None` as default and later:
```
if hh_typen is None:
    hh_typen = ["sing", "sp1ch", "sp2ch", "coup", "coup1ch", "coup2ch"]

```

Eric-Sommer replied:
Done

hmgaudecker commented:
Please keep `sing` as prefix also when children are present. `sing` / `sp`
are terrible for searching and at least I would have to look up/inner from
the context what is meant by so. We are not playing a game of keeping named
below 6 characters 😊

Also, please add an underscore (also for couples) before `[N]ch` as per
usual convention.

I would even go as far as making `0ch` explicit. But that can be debated.



Christian Zimpelmann <notifications@github.com> schrieb am Fr., 30. Okt.
2020, 11:14:

> *@ChristianZimpelmann* commented on this pull request.
> ------------------------------
>
> In gettsim/syntethic.py
> <https://github.com/iza-institute-of-labor-economics/gettsim/pull/235#discussion_r514993344>
> :
>
> > +        new_df["alter"] = pd.Series([age_adults[1], age_children[0], age_children[1]])
> +
> +    # Make sure new household members are not heads
> +    new_df["tu_vorstand"] = False
> +    # Children are in education
> +    new_df["in_ausbildung"] = new_df["kind"]
> +    # children do not have earnings
> +    new_df.loc[new_df["kind"], "bruttolohn_m"] = 0
> +    # If single earner household, the partner is assigned zero wage as well
> +    if not double_earner:
> +        new_df.loc[~new_df["kind"], "bruttolohn_m"] = 0
> +    return new_df
> +
> +
> +def gettsim_hypo_data(
> +    hh_typen=("sing", "sp1ch", "sp2ch", "coup", "coup1ch", "coup2ch"),
>
> Ok, makes sense.
>
> I would suggest hh_typen=None as default and later:
>
> if hh_typen is None:
>     hh_typen = ["sing", "sp1ch", "sp2ch", "coup", "coup1ch", "coup2ch"]
>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/iza-institute-of-labor-economics/gettsim/pull/235#discussion_r514993344>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABGV6FJX6HPL5GTJZIOUV3SNKGXTANCNFSM4SVDJ4WQ>
> .
>

Eric-Sommer commented on gettsim/syntethic.py:
```diff
@@ -0,0 +1,203 @@
+
+    hh_typen (tuple of str):
+        Allowed Household Types:
+        - 'sing' - Single, no kids
```
done

Eric-Sommer commented:
I addressed all the comments. The one thing I did not manage to not work out is how to create a unique household id. I'm not sure it's necessary though. The current approach (lines 157ff) shows the idea, but fails because even though `bruttolohn_m` might vary across households, it will not vary among children. The only other idea I had was to create a loop counter above. Could you have a look at it @ChristianZimpelmann or @MaxBlesch?

ChristianZimpelmann commented on gettsim/synthetic.py:
```diff
@@ -0,0 +1,303 @@
+    if age_children is None:
+        age_children = [3, 8]
+    # Check inputs
+    for t in hh_typen:
```
Please also check if `hh_typen` and `n_children` is entered as list/collection and give a meaningful error message if not. Especially for `n_children`, it's not necessarily intuitive that a single number doesn't work. Maybe in this case, we should convert the single int to a one element list.

ChristianZimpelmann commented on gettsim/synthetic.py:
```diff
@@ -0,0 +1,303 @@
+        columns=["hht", "nch"], data=itertools.product(hh_typen, n_children)
+    )
+
+    df["hh_typ"] = all_types["hht"].str[:4] + "_" + all_types["nch"].astype(str) + "ch"
```
Make this more verbose. I suggest something like: `single_2_child` etc.

ChristianZimpelmann commented on gettsim/synthetic.py:
```diff
@@ -0,0 +1,303 @@
+        # the combination of hh_typ and all values from heterogeneous_vars
+        # defines one synthetic household.
+        # TODO: This does not work yet however because children always have 0 income.
+        id_columns = list(heterogeneous_vars.keys())
```
That seems too complicated. I think a (global) counter is the better idea here

ChristianZimpelmann reviewed (changes requested):
Thanks for you work, @Eric-Sommer! I added a few more suggestions.

I think we can merge it soon.

ChristianZimpelmann commented:
> If this is what happens (have not checked), you are completely right @Maximilian Blömer: Data needs to be generated without any reference to GETTSIM code, everything has to be hard-coded.

Does this apply here, @Eric-Sommer?

Eric-Sommer commented:
I'd say no. I rely on data from bedarfsgemeinschaften.yaml which is not used by gettsim at all. We can move it to a separate folder of course in order to make that distinction clearer

hmgaudecker commented:
> I'd say no. I rely on data from bedarfsgemeinschaften.yaml which is not used by gettsim at all. We can move it to a separate folder of course in order to make that distinction clearer

Great. Yes, let's think about how to organise the namespace a bit more systematically, #239 and #240 are a start.

Is this PR urgent for 0.4, btw.?

ChristianZimpelmann commented:
> > I'd say no. I rely on data from bedarfsgemeinschaften.yaml which is not used by gettsim at all. We can move it to a separate folder of course in order to make that distinction clearer
> 
> Great. Yes, let's think about how to organise the namespace a bit more systematically, #239 and #240 are a start.
> 
> Is this PR urgent for 0.4, btw.?

no

Eric-Sommer commented:
I created a separate folder `synthetic_parameters` in order to separate them from the standard ones

ChristianZimpelmann commented:
Is this pr ready for another round of review, @Eric-Sommer?

Eric-Sommer commented:
> Is this pr ready for another round of review, @Eric-Sommer?

Yes. In my view, the major issue was where to store the params needed for the hypothetical data sets. I now created a new folder for them.

hmgaudecker commented:
Maybe late to the party, but just came across this one: https://declaredesign.org/r/fabricatr/

And while I am at it, not sure I linked this one before: https://www.openrg.com/reports/SynData_QN.pdf (probably out of scope of this PR, also did not read in detail whether feasible for us)

hmgaudecker commented:
And another one, this time Python: https://github.com/sdv-dev/SDV

Eric-Sommer commented on gettsim/synthetic.py:
```diff
@@ -0,0 +1,340 @@
+        "kind",
+        "rentner",
+        "betreuungskost_m",
+        "miete_unterstellt",
```
Yes. One can receive wohngeld, ALG2 and the like also if you're living in your own property.

Eric-Sommer commented on gettsim/synthetic.py:
```diff
@@ -0,0 +1,340 @@
+    df["tu_vorstand"] = True
+    df["alter"] = age_adults[0]
+    df["immobilie_baujahr"] = baujahr
+    for c in ["arbeitsl_lfdj_m", "arbeitsl_vorj_m", "arbeitsl_vor2j_m"]:
```
Totally agree

ChristianZimpelmann commented on docs/tutorials/advanced_usage.ipynb:
```diff
@@ -925,7 +921,174 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 7,
```
Shouldn't this be cleaned up?
